### PR TITLE
ImageName should be able to parse image names with IPv6 address literals

### LINKED ISSUE #2541: ImageName should be able to parse image names with IPv6 address literals
### Description
Currently, ImageName only accepts domain names or IPv4 literals in the registry name. 

However, parsing IPv6 literals support was added Docker Distrubution Reference.

We should modify our registry host regular expressions to accommodate IPv6 addresses.

Distribution's Regexp can be seen in [regexp.go](https://github.com/distribution/reference/blob/8507c7fcf0da9f570540c958ea7b972c30eeaeca/regexp.go#L91)

```golang
domainName = domainNameComponent + anyTimes(`\.`+domainNameComponent)
ipv6address = `\[(?:[a-fA-F0-9:]+)\]`

host = `(?:` + domainName + `|` + ipv6address + `)`
```

### Expected Behavior
ImageName should be able to parse image name with registry having IPv6 literal (e.g `[2001:db8::1]:5000/repo:tag`)

### Acceptance Criteria
- [ ] ImageName validation logic is modified to handle IPv6 addresses

### PR DESCRIPTION
**Pull request by Devashishbasu**
1.added IPv6 address
2.upated DOMAIN_REGEX for the IPv6 addresses
3.uncommented the related tests

Fixes #2541 

## Type of change
 - [ ] Bug fix (non-breaking change which fixes an issue)

## Checklist
 - [X ] I have read the [contributing guidelines](https://www.eclipse.dev/jkube/contributing)
 - [X ] I signed-off my commit with a user that has signed the [Eclipse Contributor Agreement](https://www.eclipse.org/legal/ECA.php)
 - [ X] My code follows the style guidelines of this project
 - [X ] I Keep It Small and Simple: The smaller the PR is, the easier it is to review and have it merged
 - [X ] I use [conventional commits](https://www.conventionalcommits.org/) in my commit messages
 - [X ] I have performed a self-review of my code

### CODE REVIEW

**Comment by manusa** [comment]
### Eclipse JKube [CI Report](https://github.com/jkubeio/ci)

Started new GH workflow run for https://github.com/eclipse/jkube/pull/2603 (_2024-02-01T05:02:19Z_)

:gear: [JKube E2E Tests (7736495585)](https://github.com/jkubeio/ci/actions/runs/7736495585) :hourglass_flowing_sand:
- :hourglass_flowing_sand: [Bootstrap CI test run (#2603)](https://github.com/jkubeio/ci/actions/runs/7736495585/job/21093943608)

File: jkube-kit/config/image/src/main/java/org/eclipse/jkube/kit/config/image/ImageName.java [hunk 1]
```diff
@@ -352,7 +353,8 @@ private boolean isRegistryValidPathComponent() {
     private static final Pattern IMAGE_NAME_REGEXP = Pattern.compile(NAME_COMPONENT_REGEXP + "(?:(?:/" + NAME_COMPONENT_REGEXP + ")+)?");
 
     // https://github.com/docker/docker/blob/04da4041757370fb6f85510c8977c5a18ddae380/vendor/github.com/docker/distribution/reference/regexp.go#L31
-    private static final Pattern DOMAIN_REGEXP = Pattern.compile("^" + DOMAIN_COMPONENT_REGEXP + "(?:\\." + DOMAIN_COMPONENT_REGEXP + ")*(?::[0-9]+)?$");
+    private static final Pattern DOMAIN_REGEXP = Pattern.compile("^(?:" + IPV6_ADDRESS_REGEXP + "|" + DOMAIN_COMPONENT_REGEXP + "(?:\\." + DOMAIN_COMPONENT_REGEXP + ")*)" +
+        "(?::\\d+)?$");
```
Comment (rohanKanojia) [inline jkube-kit/config/image/src/main/java/org/eclipse/jkube/kit/config/image/ImageName.java:357]: To make it easier to read, could you break this down into more REGEXP COMPONENTS:
```
private static final String OPTIONAL_PORT_REGEXP = "(?::[0-9]+)?";
private static final String DOMAIN_NAME_REGEXP = DOMAIN_COMPONENT_REGEXP + "(?:\\." + DOMAIN_COMPONENT_REGEXP + ")*";
private static final String REGISTRY_HOST_REGEXP = "^(?:" + IPV6_ADDRESS_REGEXP + "|" + DOMAIN_NAME_REGEXP + ")";

private static final Pattern REGISTRY_REGEXP = Pattern.compile(REGISTRY_HOST_REGEXP + OPTIONAL_PORT_REGEXP);
```

File: jkube-kit/config/image/src/main/java/org/eclipse/jkube/kit/config/image/ImageName.java [hunk 2]
```diff
@@ -352,7 +353,8 @@ private boolean isRegistryValidPathComponent() {
     private static final Pattern IMAGE_NAME_REGEXP = Pattern.compile(NAME_COMPONENT_REGEXP + "(?:(?:/" + NAME_COMPONENT_REGEXP + ")+)?");
 
     // https://github.com/docker/docker/blob/04da4041757370fb6f85510c8977c5a18ddae380/vendor/github.com/docker/distribution/reference/regexp.go#L31
```
Comment (rohanKanojia) [inline jkube-kit/config/image/src/main/java/org/eclipse/jkube/kit/config/image/ImageName.java:355]: Could you please update the comment for regexp source too?

File: jkube-kit/config/image/src/main/java/org/eclipse/jkube/kit/config/image/ImageName.java (hunk 1 above)
Comment (Devashishbasu) [reply jkube-kit/config/image/src/main/java/org/eclipse/jkube/kit/config/image/ImageName.java]: > To make it easier to read, could you break this down into more REGEXP COMPONENTS:
> 
> ```
> private static final String OPTIONAL_PORT_REGEXP = "(?::[0-9]+)?";
> private static final String DOMAIN_NAME_REGEXP = DOMAIN_COMPONENT_REGEXP + "(?:\\." + DOMAIN_COMPONENT_REGEXP + ")*";
> private static final String REGISTRY_HOST_REGEXP = "^(?:" + IPV6_ADDRESS_REGEXP + "|" + DOMAIN_NAME_REGEXP + ")";
> 
> private static final Pattern REGISTRY_REGEXP = Pattern.compile(REGISTRY_HOST_REGEXP + OPTIONAL_PORT_REGEXP);
> ```

Done

**Comment by Devashishbasu** [comment]
Hii @rohanKanojia sonarcloud asking to refactor this code should i do it?
image

**Comment by rohanKanojia** [comment]
@Devashishbasu : Could you please check Docker's [distribution/regexp.go](https://github.com/distribution/reference/blob/8507c7fcf0da9f570540c958ea7b972c30eeaeca/regexp.go) if they also have same repetition. We should be aligned with what's in docker source.

**Comment by Devashishbasu** [comment]
> @Devashishbasu : Could you please check Docker's [distribution/regexp.go](https://github.com/distribution/reference/blob/8507c7fcf0da9f570540c958ea7b972c30eeaeca/regexp.go) if they also have same repetition. We should be aligned with what's in docker source.



> Member

I have checked it with Docker's [distribution/regexp.go](https://github.com/distribution/reference/blob/8507c7fcf0da9f570540c958ea7b972c30eeaeca/regexp.go) and with master branch also...the new expression is aligned with Docker's [distribution/regexp.go] (https://github.com/distribution/reference/blob/8507c7fcf0da9f570540c958ea7b972c30eeaeca/regexp.go)     and the sonar cloud is also suggesting to refactor optional port regex 
image

what should we do now?

**Comment by rohanKanojia** [comment]
@Devashishbasu : You can ignore the sonar warning.

**Comment by Devashishbasu** [comment]
> > @Devashishbasu : You can ignore the sonar warning.
> 
> So is it ready to merge? or I have do anything else?

Hii @rohanKanojia

File: jkube-kit/config/image/src/main/java/org/eclipse/jkube/kit/config/image/ImageName.java [hunk 3]
```diff
@@ -351,8 +354,13 @@ private boolean isRegistryValidPathComponent() {
     // https://github.com/docker/docker/blob/04da4041757370fb6f85510c8977c5a18ddae380/vendor/github.com/docker/distribution/reference/regexp.go#L53
     private static final Pattern IMAGE_NAME_REGEXP = Pattern.compile(NAME_COMPONENT_REGEXP + "(?:(?:/" + NAME_COMPONENT_REGEXP + ")+)?");
 
+    // https://github.com/distribution/reference/blob/8507c7fcf0da9f570540c958ea7b972c30eeaeca/regexp.go#L65
+    private static final String OPTIONAL_PORT_REGEXP = "(?::[0-9]+)?";
+    private static final String DOMAIN_NAME_REGEXP = DOMAIN_COMPONENT_REGEXP + "(?:\\." + DOMAIN_COMPONENT_REGEXP + ")*";
```
Comment (rohanKanojia) [inline jkube-kit/config/image/src/main/java/org/eclipse/jkube/kit/config/image/ImageName.java:359]:
```suggestion
   // https://github.com/distribution/reference/blob/8507c7fcf0da9f570540c958ea7b972c30eeaeca/regexp.go#L99
    private static final String DOMAIN_NAME_REGEXP = DOMAIN_COMPONENT_REGEXP + "(?:\\." + DOMAIN_COMPONENT_REGEXP + ")*";
```

File: jkube-kit/config/image/src/main/java/org/eclipse/jkube/kit/config/image/ImageName.java [hunk 4]
```diff
@@ -351,8 +354,13 @@ private boolean isRegistryValidPathComponent() {
...
+    private static final String REGISTRY_HOST_REGEXP = "^(?:" + IPV6_ADDRESS_REGEXP + "|" + DOMAIN_NAME_REGEXP + ")";
```
Comment (rohanKanojia) [inline jkube-kit/config/image/src/main/java/org/eclipse/jkube/kit/config/image/ImageName.java:360]:
```suggestion
    // https://github.com/distribution/reference/blob/8507c7fcf0da9f570540c958ea7b972c30eeaeca/regexp.go#L106
    private static final String REGISTRY_HOST_REGEXP = "^(?:" + IPV6_ADDRESS_REGEXP + "|" + DOMAIN_NAME_REGEXP + ")";
```

File: jkube-kit/config/image/src/main/java/org/eclipse/jkube/kit/config/image/ImageName.java [hunk 5]
```diff
@@ -351,8 +354,13 @@ private boolean isRegistryValidPathComponent() {
...
+
     // https://github.com/docker/docker/blob/04da4041757370fb6f85510c8977c5a18ddae380/vendor/github.com/docker/distribution/reference/regexp.go#L31
-    private static final Pattern DOMAIN_REGEXP = Pattern.compile("^" + DOMAIN_COMPONENT_REGEXP + "(?:\\." + DOMAIN_COMPONENT_REGEXP + ")*(?::[0-9]+)?$");
+    private static final Pattern REGISTRY_REGEXP = Pattern.compile(REGISTRY_HOST_REGEXP + OPTIONAL_PORT_REGEXP);
```
Comment (rohanKanojia) [inline jkube-kit/config/image/src/main/java/org/eclipse/jkube/kit/config/image/ImageName.java:363]:
```suggestion
    // https://github.com/distribution/reference/blob/8507c7fcf0da9f570540c958ea7b972c30eeaeca/regexp.go#L110
    private static final Pattern REGISTRY_REGEXP = Pattern.compile(REGISTRY_HOST_REGEXP + OPTIONAL_PORT_REGEXP);
```

**Comment by rohanKanojia** [comment]
Looks good, only minior comment regarding adding regexp sources

**Pull request closed without merging**